File: ScsiServer/scsiserver.c
#include <sys/types.h>	/* for socket, bind, htonl, htons, fork, waitpid */
#include <sys/socket.h>	/* for socket, bind */
#include <netdb.h>	/* for getprotobyname */
#include <netinet/in.h>	/* for htonl, htons */
#include <unistd.h>	/* for fork */
#include <stdlib.h>	/* for malloc */
#include <signal.h>	/* for signal */
#include <sys/wait.h>	/* for waitpid */
#include <errno.h>	/* for errno, EINTR */
#include <stdio.h>	/* for stderr */

#include <strings.h>
#include "ScsiTransport.h"


#define PORT 39074
#define QUEUESIZE 10


typedef unsigned char byte;


int use_children = 1;


static int
myread(int fd, void *buf, int nbyte)
/* try harder to read nbytes into buf from fd */
/* only stop when there's nothing left to read or nbyte has been satisfied */
/* return number of bytes read */
{
  int curread;
  int bytesleft = nbyte;
  while (bytesleft) {
    curread = read(fd, buf, bytesleft);
    if (curread == 0)
      break;
    if (curread == -1)
      return -1;
    buf       += curread;
    bytesleft -= curread;
  }
  return nbyte-bytesleft;
}


void *
mymalloc(size_t size, char *errstring)
{
  void *retval;
  if (size) {
    retval = malloc(size);
    if (retval == NULL) {
      /* this should probably be a syslog instead of fprintf */
      fprintf(stderr, "malloc failed '%s'\n", errstring);
      exit(1);
    }
  } else {
    retval = NULL;
  }
  return retval;
}


void one_connection(int fd)
{
  for (;;) {
    byte devlen;
    byte *dev;
    SCSI_HANDLE handle;
    DIRECTION rw;
    byte cdblen;
    byte cdb[16];	/* assuming well-behaved client */
    int  datlen;
    byte *dat = NULL;
    int  sttlen;
    byte *stt;
    int  timout;
    byte status;
    long q;
    byte qc;

    if (!myread(fd, &devlen, 1)     ) break;
    dev = mymalloc(devlen+1, "dev");
    if (!myread(fd, dev    , devlen)) break; dev[devlen] = '\0';
    if (!myread(fd, &qc    , 1)     ) break;
    rw = (DIRECTION)qc;
    if (!myread(fd, &cdblen, 1)     ) break;
    if (!myread(fd, cdb    , cdblen)) break;
    if (!myread(fd, &datlen, 4)     ) break; datlen = ntohl(datlen);
    switch ((char)rw) {
      case DIRECTION_NONE: break;
      case DIRECTION_OUT:  break;
      case DIRECTION_IN:   break;
      case 'r': case 'R': rw = (datlen==0)?DIRECTION_NONE:DIRECTION_IN ; break;
      case 'w': case 'W': rw = (datlen==0)?DIRECTION_NONE:DIRECTION_OUT; break;
    }
    dat = mymalloc(datlen, "dat");
    if (rw == DIRECTION_OUT && dat != NULL) {
      if (!myread(fd, dat    , datlen)) break;
    }
    if (!myread(fd, &sttlen, 4)     ) break; sttlen = ntohl(sttlen);
    if (!myread(fd, &timout, 4)     ) break; timout = ntohl(timout);
    stt = mymalloc(sttlen, "stt");
    scsi_open(&handle, dev);
    status = (handle->cdb)(handle, rw, cdb, cdblen, dat, &datlen, stt, &sttlen, (float)timout);
    (handle->close)(&handle);
    /*debug*/fprintf(stderr, "server: datlen = %d\n", datlen);/**/
    if (rw != DIRECTION_IN) datlen = 0;
    write(fd, &status, 1);
    q = htonl(datlen);
    write(fd, &q     , 4); write(fd, dat, datlen);
    q = htonl(sttlen);
    write(fd, &q     , 4); write(fd, stt, sttlen);
    free(stt);
    free(dat);
    free(dev);
  }
  close(fd);
}


void
child_death(int p)
{
  waitpid(-1, NULL, WNOHANG);
  signal(SIGCHLD, child_death);
}


int
main()
{
  int S, NS;
  struct sockaddr_in srv;
  struct sockaddr_in addr;
  socklen_t addr_len = sizeof(addr);

  /* S = socket(AF_INET, SOCK_STREAM, (getprotobyname("tcp"))->p_proto); */
  S = socket(AF_INET, SOCK_STREAM, 0);
  if (S < 0) {
    perror("socket");
    exit(1);
  }

  srv.sin_family = AF_INET;
  srv.sin_addr.s_addr = htonl(INADDR_ANY);
  srv.sin_port = htons(PORT);
  if (bind(S, (struct sockaddr *) &srv, sizeof(srv)) < 0) {
    perror("bind");
    exit(1);
  }

  if (listen(S, QUEUESIZE) < 0) {
    perror("listen");
    exit(1);
  }

  signal(SIGCHLD, child_death);

  for (;;) {
    pid_t pid;

    for (;;) {
      NS = accept(S, (struct sockaddr *)&addr, &addr_len);
      if (NS >= 0) break;
      if (errno != EINTR) {
	perror("accept");
	exit(1);
      }
    }

    if (use_children) {
      if ((pid = fork())) {
	/* parent */
	close(NS);
      } else if (pid == (pid_t)-1) {
	/* fork failed */
	perror("fork");
	exit(1);
      } else {
	/* child */
	one_connection(NS);
	exit(0);
      }
    } else {
      one_connection(NS);
    }
  }

}

File: old/scsiserver.pl
#!/usr/bin/perl -w
# -*- mode: perl -*-
# $Header$

#push(@INC, "/tmp");
#push(@INC, "/usr/lib/perl");
#push(@INC, "/usr/local/lib/perl");
#push(@INC, "/usr/ih/bin");
#push(@INC, "/usr/local/bin");
#unshift(@INC, "/manuf/manuf/bin");

use SCSImod;

$use_children = 1;
$sockaddr = "S n a4 x8";
$AF_INET = 2;
$SOCK_STREAM = 2;
$port = 39074;

# to make -w happy
$name = "";
$alias = "";
$addr = "";

($name, $alias, $proto) = getprotobyname('tcp');
$this = pack($sockaddr, $AF_INET, $port, "\0\0\0\0");

socket(S, $AF_INET, $SOCK_STREAM, $proto) || die "socket: $!";
bind  (S, $this) || die "bind: $!";
listen(S,50) || die "listen: $!";

select((select(S), $| = 1)[0]);

$fd=$port;
for (;;) {
    $fd++;
    ($addr = accept($fd, S)) || die $!;
    if ($use_children) {
	if ($pid = fork()) {
	    # parent
	    close($fd);
	} elsif (defined $pid) {
	    # child
	    &one_connection($fd);
	    exit(0);
	} else {
	    die "Can't fork: $!\n";
	}
    } else {
	&one_connection($fd);
    }
}


sub one_connection {
    my($fd) = @_;
    my($devlen, $dev, $rw, $cdblen, $cdb, $datlen, $dat, $sttlen, $timout) =
	("","","","","","","","","");
    select((select($fd), $| = 1)[0]);
    for (;;) {
	&myread($fd, $devlen, 1)   || last; $devlen = unpack("C", $devlen);
	&myread($fd, $dev, $devlen)|| last;
	&myread($fd, $rw, 1)       || last;
	&myread($fd, $cdblen, 1)   || last; $cdblen = unpack("C", $cdblen);
	&myread($fd, $cdb, $cdblen)|| last;
	&myread($fd, $datlen, 4)   || last; $datlen = unpack("N", $datlen);
	if ($rw eq "w") {
	    &myread($fd, $dat, $datlen)|| last;
	}
	&myread($fd, $sttlen, 4)   || last; $sttlen = unpack("N", $sttlen);
	&myread($fd, $timout, 4)   || last; $timout = unpack("N", $timout);
	$status = SCSImod::send_cdb(
				    $dev,
				    $rw,
				    $cdb,
				    $dat, $datlen,
				    $sttlen ? $stt : undef, $sttlen,
				    $timout,
				    );
	$sttlen = length($stt);
	if ($rw ne "r") { $dat = ""; }
	$datlen = length($dat);
	print $fd
	    pack("C", $status),
	    pack("N", $datlen), $datlen ? $dat : "",
	    pack("N", $sttlen), $stt;
    }
    close($fd);
}


sub myread {
    # this "my" is translated into "local" for perl4 on suns
    # this illustrates the problem with "local": I can't use "datlen" here
    # because I'm writing into it as $_[$datref]
    my($fd, $mydat, $mydatlen) = @_;
    my($datref) = (1);
    my($retval);
    $_[$datref] = $mydat = "";
    while (length($_[$datref]) < $mydatlen) {
	$retval = read($fd, $_[$datref], $mydatlen-length($_[$datref]), length($_[$datref]));
	last if (!defined($retval) || $retval == 0);
    }
    return defined $retval ? length($_[$datref]) : $retval;
}

